File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        ans = 0 
        n = len(heights) 
        stack =[] 
        leftMost = [-1] *n
        rightMost = [n] * n

        for i in range(n):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()
            if stack:
                leftMost[i] = stack[-1]
            stack.append(i)

        stack = []  

        for i in range(n-1, -1, -1):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()
            if stack:
                rightMost[i] = stack[-1]
            stack.append(i)

        for i in range(n):
            leftMost[i] += 1 
            rightMost[i] -= 1
            area = (rightMost[i] - leftMost[i] + 1) * heights[i]
            ans = max(ans, area)
        return ans
                    

File: Data Structures & Algorithms/max-water-container/submission-1.cs
public class Solution {
    public int MaxArea(int[] heights) {
        int ans = 0 ; 
        int l = 0 , r = heights.Length -1 ;
        while(l < r){
            int lv = heights[l] ;
            int rv = heights[r] ;
            int min = Math.Min(lv, rv) ;

            ans = Math.Max(ans, min * (r-l));

            if(lv < rv) {
                l++;
            }else{
                r--;
            }

        }
        

        return ans ;

    }
}
